test/integration/main.cpp

```diff
@@ -2094,4 +2103,197 @@ TEST_F(ZNCTest, ModuleCrypt) {
+    client2.Write("USER user2/test x x :x");
+    Z;
+    client2.Write("znc loadmod otr");
+    client2.ReadUntil("Loaded module");
+    Z;
+
+    // genkey without --really should produce a warning
+    client1.Write("PRIVMSG *otr :genkey");
+    client1.ReadUntil("genkey --really");
+    Z;
+
+    // generate keys
+    client1.Write("PRIVMSG *otr :genkey --really");
+    client1.ReadUntil("Key generation finished.");
+    Z;
+    client2.Write("PRIVMSG *otr :genkey --really");
+    client2.ReadUntil("Key generation finished.");
+    Z;
+
+    // do not overwrite existing key without --overwrite
+    client2.Write("PRIVMSG *otr :genkey --really");
+    client2.ReadUntil("genkey --overwrite");
+    Z;
+
+    // OTR handshake
+    client1.Write("PRIVMSG nick2 :?OTR?");
+    ircd1.RelayPrivMsg(ircd2, ":user!user@user/test ", IsOtrCiphertext);
+    Z;
+    ircd2.RelayPrivMsg(ircd1, ":nick2!user2@user2/test ", IsOtrCiphertext);
+    Z;
+    ircd1.RelayPrivMsg(ircd2, ":user!user@user/test ", IsOtrCiphertext);
+    Z;
+    ircd2.RelayPrivMsg(ircd1, ":nick2!user2@user2/test ", IsOtrCiphertext);
+    Z;
+    ircd2.RelayPrivMsg(ircd1, ":nick2!user2@user2/test ", IsOtrCiphertext);
+    Z;
+    ircd1.RelayPrivMsg(ircd2, ":user!user@user/test ", IsOtrCiphertext);
+    Z;
+    ircd1.RelayPrivMsg(ircd2, ":user!user@user/test ", IsOtrCiphertext);
+    Z;
+    client1.ReadUntil("Peer is not authenticated");
+    client2.ReadUntil("Peer is not authenticated");
+    Z;
+
+    // keys are initially not trusted
+    client1.Write("PRIVMSG *otr :info");
+    client1.ReadUntil("not trusted |");
+    Z;
+    client2.Write("PRIVMSG *otr :info");
+    client2.ReadUntil("not trusted |");
+    Z;
+
+    // send encrypted message
+    client1.Write("PRIVMSG nick2 :let's authenticate with question: what is my favorite book?");
+    ircd1.RelayPrivMsg(ircd2, ":user!user@user/test ", IsOtrCiphertext);
+    ircd1.RelayPrivMsg(ircd2, ":user!user@user/test ", IsOtrCiphertext);
+    client2.ReadUntil("let's authenticate with question: what is my favorite book?");
+    Z;
+    client2.Write("PRIVMSG user :m'kay");
+    ircd2.RelayPrivMsg(ircd1, ":nick2!user2@user2/test ", IsOtrCiphertext);
+    client1.ReadUntil("m'kay");
+    Z;
+
+    // try shared secret auth with non-matching secrets and check that it failed
+    client1.Write("PRIVMSG *otr :auth nick2 the man who was thursday");
+    Z;
+    for (int i = 0; i < 5; i++) {
+        ircd1.RelayPrivMsg(ircd2, ":user!user@user/test ", IsOtrCiphertext);
+        Z;
+    }
+    client2.ReadUntil("To complete, type auth user <secret>.");
+    Z;
+    client2.Write("PRIVMSG *otr :auth user hunter2");
+    Z;
+    for (int i = 0; i < 8; i++) {
+        ircd2.RelayPrivMsg(ircd1, ":nick2!user2@user2/test ", IsOtrCiphertext);
+        Z;
+    }
+    for (int i = 0; i < 6; i++) {
+        ircd1.RelayPrivMsg(ircd2, ":user!user@user/test ", IsOtrCiphertext);
+        Z;
+    }
+    for (int i = 0; i < 3; i++) {
```

Comment:
That's the number of lines/messages sent in each direction during the shared authentication. Same problem as above.